### PR TITLE
fix https://github.com/dmlc/gluon-nlp/issues/1315

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description ##
For each worker, we should normalize the loss by the number of elements to improve numerical stability and can avoid setting `step_size` in trainer.update(), and this is consistent with the official TF implementation, see https://github.com/google-research/electra/blob/master/run_pretraining.py#L181 and https://github.com/google-research/electra/blob/master/run_pretraining.py#L206.

To check the correctness of the Horovod support, let k denote the number of workers,

Ground truth:
```
g = \frac{1}{k} \sum_{i=1}{k} g_i

update_gt = clip_grad_global_norm(g, max_norm)
```

In our code
```
G = \sum_{i=1}{k} g_i     (by trainer.allreduce_grads())

update = \frac{1}{k} * clip_grad_global_norm(G, max_norm * k)

         (note that \frac{1}{k} is the scaling factor inside hvd.trainer)

       = \frac{1}{k} * clip_grad_global_norm(g * k, max_norm * k)

       = \frac{1}{k} * k * clip_grad_global_norm(g, max_norm)

       = update_gt
```
And this part of implementation for Horovod support is consistent with BERT script, see https://github.com/dmlc/gluon-nlp/blob/v0.10.x/scripts/bert/run_pretraining.py.

@sxjscience  Could you also check the correctness?

Also, we need re-train the model and adjust the hyper-parameters accordingly.

### Essentials ###
- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

cc @dmlc/gluon-nlp-team

### CODE REVIEW

File: scripts/pretraining/run_electra.py [hunk 1]
```diff
@@ -378,20 +377,22 @@ def train(args):
                 with mx.autograd.record():
                     mlm_scores, rtd_scores, corrupted_tokens, labels = model(
                         masked_input_ids, segment_ids, valid_lengths, unmasked_tokens, masked_positions)
-                    # the official implementation takes the sum of each batch inside the loss function
-                    # while SigmoidBinaryCrossEntropyLoss and SoftmaxCELoss takes the mean value
+                    num_elements = masked_weights.sum()
                     mlm_loss = mlm_loss_fn(
-                        mlm_scores, unmasked_tokens, masked_weights.reshape(-1)).mean() / (masked_weights.mean() + 1e-6)
+                        mx.nd.reshape(mlm_scores.as_nd_ndarray(), (-3, -1)).as_np_ndarray(),
```
Comment (sxjscience) [inline scripts/pretraining/run_electra.py:382]: For advanced reshape, you can use `mx.npx.reshape(mlm_scores, (-4, -2))`

**Comment by sxjscience** [comment]
LGTM. Need to revise the usage of `nd.reshape` --> `npx.reshape`.

File: scripts/pretraining/run_electra.py [hunk 2]
```diff
@@ -402,24 +403,28 @@ def train(args):
             log_rtd_loss += sum([ele.as_in_ctx(ctx_l[0])
                                  for ele in rtd_loss_l]).asnumpy()
             log_total_loss += sum([ele.as_in_ctx(ctx_l[0])
-                                   for ele in loss_l]).asnumpy() * loss_denom
+                                   for ele in loss_l]).asnumpy()
```
Comment (ZheyuYe) [inline scripts/pretraining/run_electra.py:406]: Does `log_total_loss` need to multiply by this coefficient `args.num_accumulated` here?

Comment (sxjscience) [reply scripts/pretraining/run_electra.py:406]: I think each loss was divided by `args.num_accumulated` so there is no need to multiply `args.num_accumulated`

Comment (ZiyueHuang) [reply scripts/pretraining/run_electra.py:406]: @sxjscience `log_total_loss` is only used in logging.info(). And I notice that the logging phase also has some problems (but irrelevant to the training), I will fix this later.

Comment (ZheyuYe) [reply scripts/pretraining/run_electra.py:406]: During the log recording phase, is `log_total_loss` equal to the weighted sum of `log_rtd_loss` and `log_mlm_loss` for now?

**Comment by ZiyueHuang** [comment]
I've made some change. Could you please review again?
